### PR TITLE
Generate "tell" variants for functions not returning anything in the stub generator (WIP)

### PR DESCRIPTION
**Pull request by afsalthaj**
Fixing https://github.com/golemcloud/golem/issues/278

Drafting the PR to get earlier review

I couldn’t change the wit generation and proceed, because the source generation depends on the original stub definitions.
Therefore, just changed at the site of collection function stub.

### CODE REVIEW

**Comment by afsalthaj** [comment]
image


image


image

File: wasm-rpc-stubgen/src/stub.rs [hunk 1]
```diff
@@ -332,37 +332,47 @@ fn collect_stub_interfaces(resolve: &Resolve, world: &World) -> anyhow::Result<V
...
-        .collect())
+            Results::Anon(single) => FunctionResultStub::Single(*single),
+        };
+
+        if results.is_empty() {
+            function_stubs.push(FunctionStub {
+                name: format!("blocking-{}", f.name),
+                params: params.clone(),
+                results: results.clone(),
+            })
```
Comment (justcoon) [inline wasm-rpc-stubgen/src/stub.rs:365]: if results is empty, then there will be 2 functions ?

1. blocking-function_name
2. function_name 
?

**Review by justcoon — APPROVED** [review]

**Review by vigoo — CHANGES_REQUESTED** [review]
There are a couple of problems. 

Most importantly, you are sending a wrong function name to the `invoke-and-await` function - there is no `blocking-add-event` function exported on the remote worker, just `add-event`.

It also does not look nice that you generate both functions into the same function collection and then make decisions later based on their name (`starts_with`).

I suggest to:
- do not modify `stub.rs` - keep the list of functions as it is
- in `rust.rs` modify `generate_function_stub_source` so it gets a parameter that controls its behavior - based on that it can decide whether to override the function name (but not in the invoke's string parameter!) or not, and whether to use the `invoke` or the `invoke-and-await`. Then you need to call this twice in the `generate_stub_source` if the function has no return values, to generate both variants.
- Do the same in `wit.rs` - just based on the `FunctionStub` you can see if you need to generate two functions or one.
- The whole thing has to work for global functions, resource methods, and static functions.

**Comment by afsalthaj** [comment]
@vigoo  Done

**Comment by afsalthaj** [comment]
image

**Review by vigoo — APPROVED** [review]

**Pull request merged**